src/integrationtests/java/com/aws/greengrass/integrationtests/deployment/DeploymentTaskIntegrationTest.java

```diff
@@ -1246,6 +1246,38 @@ void GIVEN_services_running_WHEN_new_deployment_asks_to_skip_update_policy_check
         assertEquals(DeploymentResult.DeploymentStatus.SUCCESSFUL, result.getDeploymentStatus());
     }
 
+    @Test
+    @Order(101)
+    void GIVEN_services_with_dependency_broken_WHEN_new_deployment_failure_handling_policy_rollback_THEN_deployment_rollback_is_successful(
+            ExtensionContext context) throws Exception {
+
+        ignoreExceptionUltimateCauseOfType(context, ServiceUpdateException.class);
+
+        // Deploy a broken config with no rollback
+        Future<DeploymentResult> resultFuture = submitSampleJobDocument(
+                DeploymentTaskIntegrationTest.class.getResource("DependencyDoNothing.json").toURI(),
+                System.currentTimeMillis());
+        DeploymentResult result = resultFuture.get(DEPLOYMENT_TIMEOUT, TimeUnit.SECONDS);
+        resultFuture.get(60, TimeUnit.SECONDS);
+        List<String> services = kernel.orderedDependencies().stream()
+                .filter(greengrassService -> greengrassService instanceof GenericExternalService)
+                .map(GreengrassService::getName).collect(Collectors.toList());
+        // should contain main, Nucleus, BreakingService2, DependencyOnBreak
+        assertThat(services, containsInAnyOrder("main", DEFAULT_NUCLEUS_COMPONENT_NAME,
+                "BreakingService2", "DependencyOnBreak", "DependencyOnDependency", "SoftDependencyOnBreak"));
+        assertEquals("1.0.0", kernel.findServiceTopic("DependencyOnBreak").find("version").getOnce());
+        assertEquals(DeploymentResult.DeploymentStatus.FAILED_ROLLBACK_NOT_REQUESTED, result.getDeploymentStatus());
+        preloadLocalStoreContent();
+
+        // Deploy a new config, increase depender version, with rollback.
+        resultFuture =
+                submitSampleJobDocument(DeploymentTaskIntegrationTest.class.getResource("DependencyRollback.json").toURI(),
+                        System.currentTimeMillis());
+        result = resultFuture.get(DEPLOYMENT_TIMEOUT, TimeUnit.SECONDS);
+        assertEquals("1.0.0", kernel.findServiceTopic("DependencyOnBreak").find("version").getOnce());
+        assertEquals(DeploymentResult.DeploymentStatus.FAILED_ROLLBACK_COMPLETE, result.getDeploymentStatus());
+    }
```

Comment:
can you also check that an empty deployment can remove all these components?

Reply:
Updated